csrc/device_lower/analysis/tma.cpp

```diff
@@ -378,105 +375,96 @@ TMAInfo getTMAInfo(LoadStoreOp* ldst) {
 
   // So far, we have infered the TMA domain. The size of TMA domain is not
   // necessarily the dimensionality of TMA because we support defining box
-  // by compositing. We use a state machine to infer the dimensions of TMA.
-  //
+  // by compositing. We use AbstractTensor to further merge the TMA domain to
```

Comment:
IIUC, we are creating the true allocation domain dynamically here. If so, why not do so when defining the tensor, not just doing here dynamically? If the tensor has the right allocation domain as defined below, I suppose the indexing should just work.

Reply:
To be clear, I'm not asking for a change. Just want to make sure if what I'm thinking about makes any sense.

Reply:
Yes, we are creating the true allocation domain dynamically here. And you are correct that, if we just define it in the tensor, it will work.

The reason for not defining it in the tensor is exactly the same reason of why supporting this "define box by compositing" at all. It is mostly about "There is this labor that I have to do, should I do it in the scheduler, or should I do it in lowering?" and I choose to do it in lowering. The reason for this decision is because scheduling is already a very difficult problem, so I want to free the scheduler from having to do the book keeping work of TMA.

For example, if we have a tensor `[I0, I1, I2, I3, I4, I5]`, and I want to schedule a pointwise relu kernel. With this "define box by compositing" mechanism, the scheduler author could just think like below:

> I will just tile the tensor on the innermost two dims as `(4, 256)` and the other dims will be automatically handled.

However, without this "define box by compositing" mechanism, the scheduler author still need to analyze the contiguity of the input tensor, and merge contiguous dims together to set the allocation domain of the global tensor, so that we will not use a non-existing 6D TMA. Also, non-permute allocation domains on fusion inputs/outputs could be hard for things like, stride order inference.

Reply:
Thanks. That makes sense.